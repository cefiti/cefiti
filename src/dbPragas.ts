//CEFiTI - Pragas

interface Files {
  leg: string
  link: string
}

export interface Praga {
  prag: string
  hosp: string[]
  pragc: string
  files: Files[]
}

const pragas: Praga[] = [
  {
    prag: 'Bactrocera carambolae',
    pragc: 'MOSCA-DA-CARAMBOLA',
    hosp: [
      'Anacardium occidentale',
      'Arenga pinnata',
      'Artocarpus altilis',
      'Artocarpus integrifolia',
      'Averrhoa bilimbi',
      'Averrhoa carambola',
      'Capsicum annuum',
      'Chrysobalanus icaco',
      'Citrus aurantium',
      'Citrus paradisi',
      'Citrus reticulata',
      'Citrus sinensis',
      'Eugenia uniflora',
      'Garcinia dulcis',
      'Licopersicum esculentum',
      'Malpighia emarginata',
      'Mangifera indica',
      'Manilkara zapota',
      'Psidium guajava',
      'Spondias mombin',
      'Spondias lutea',
      'Syzygium aqueum',
      'Syzygium jambos',
      'Syzygium malaccense',
      'Syzygium samarangense',
      'Terminalia catappa',
      'Ziziphus jujuba',
      'Ziziphus mauritiana',
      'Bysonima crassifolia',
      'Capsicum chinense',
      'Chrysophyllum cainito',
      'Eugenia stipitata',
      'Licania sp.',
      'Pouteria caimito',
      'Pouteria macrophylla',
      'Psidium guineense',
      'Rollinia omucosa',
      'Syzygium cumini',
    ],
    files: [
      { leg: 'Instrução Normativa nº 28, de 20 de julho de 2017', link: 'IN28-2017.pdf' },
      { leg: 'Resulução nº 03, de 15 de maio de 2018', link: 'RES03-2018.pdf' },
      { leg: 'Resulução nº 06, de 23 de novembro de 2018', link: 'RES06-2018.pdf' },
    ],
  },
  {
    prag: 'Schizotetranychus hindustanicus',
    pragc: 'ÁCARO HINDU',
    hosp: [
      'Cocos nucifera',
      'Azadirachta indica',
      'Sorghum bicolor',
      'Acacia spp.',
      'Melia azedarach',
      'Citrus spp.',
    ],
    files: [
      { leg: 'Instrução Normativa MAPA nº 8, de 17 de abril de 2012', link: 'IN08-2012.pdf' },
    ],
  },
  {
    prag: 'Sternochetus mangiferae',
    pragc: 'GORGULHO DA MANGA',
    hosp: ['Mangifera indica'],
    files: [
      {
        leg: 'Instrução Normativa MAPA/SDA nº. 34, de 05 de setembro de 2017',
        link: 'IN34-2017.pdf',
      },
    ],
  },
  {
    prag: 'BSV e CMV',
    pragc: 'BANANA STREAK VIRUS e CUCUMBER MOSAIC VIRUS',
    hosp: ['Musa spp.'],
    files: [
      { leg: 'Instrução Normativa MAPA nº 46, de 27 de dezembro de 2010', link: 'IN46-2010.pdf' },
    ],
  },
  {
    prag: 'Xanthomonas campestris pv. viticola',
    pragc: 'CANCRO-DA-VIDEIRA',
    hosp: ['Vitis spp. e seus híbridos'],
    files: [
      { leg: 'Instrução Normativa MAPA nº 2, de 6 de fevereiro de 2014', link: 'IN02-2014.pdf' },
    ],
  },
  {
    prag: 'Neonectria ditissima (Neonectria galligena)',
    pragc: 'CANCRO EUROPEU DAS POMÁCEAS',
    hosp: ['Malus sp.'],
    files: [{ leg: 'Instrução Normativa nº 20, de 20 de junho de 2013', link: 'IN20-2013.pdf' }],
  },
  {
    prag: 'Candidatus liberibacter americanus e Candidatus liberibacter asiaticus',
    pragc: 'GREENING',
    hosp: ['Citrus spp.', 'Fortunella spp.', 'Poncirus spp.', 'Murraya paniculata'],
    files: [{ leg: 'Instrução Normativa nº 53, de 16 de outubro de 2008', link: 'IN53-2008.pdf' }],
  },
  {
    prag: 'Phyllosticta citricarpa (Guinardia citricarpa)',
    pragc: 'PINTA-PRETA-DOS-CITROS',
    hosp: ['Citrus spp.'],
    files: [
      { leg: 'Instrução Normativa MAPA nº 3, de 8 de janeiro de 2008', link: 'IN03-2008.pdf' },
    ],
  },
  {
    prag: 'Ralstonia solanacearum raça 2',
    pragc: 'MOKO-DA-BANANEIRA',
    hosp: ['Musa spp.', 'Heliconia spp.'],
    files: [
      { leg: 'Instrução Normativa SDA/MAPA nº 17, de 27 de maio de 2009', link: 'IN17-2009.pdf' },
    ],
  },
  {
    prag: 'Pseudocercospora fijiensis (Mycosphaerella fijiensis)',
    pragc: 'SIGATOKA NEGRA',
    hosp: ['Musa spp.', 'Heliconia spp.'],
    files: [
      { leg: 'Instrução Normativa SDA/MAPA nº 17, de 31 de maio de 2005', link: 'IN17-2005.pdf' },
    ],
  },
  {
    prag: 'Anthonomus tomentosus',
    pragc: 'ANTHONOMUS TOMENTOSUS',
    hosp: ['Malpighia spp.'],
    files: [
      {
        leg: 'Instrução Normativa MAPA/SDA nº. 38, de 01 de outubro de 2018',
        link: 'IN38-2018.pdf',
      },
    ],
  },
  {
    prag: 'Xanthomonas citri subsp. citri',
    pragc: 'CANCRO CÍTRICO',
    hosp: ['Citrus spp.', 'Fortunella spp.', 'Poncirus spp.'],
    files: [{ leg: 'Instrução Normativa nº 21, de 25 de abril de 2018', link: 'IN21-2018.pdf' }],
  },
]

export { pragas }
