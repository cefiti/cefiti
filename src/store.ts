import { observable, computed, configure, action, runInAction } from 'mobx'
import 'js-plus'

configure({ enforceActions: 'observed' })

export class Store {
  @observable dbRegras: Regra[] = []
  @observable dbHospedeiros: Hospedeiro[] = []
  @observable dbPragas: Praga[] = []
  @observable db: Db[] = []
  @observable estados: Estado[] = []

  @observable
  dados = { hospSci: '', hospVul: '', prod: '', orig: '', dest: '' }

  async getDb() {
    const { regras, pragas, hospedeiros, estados } = await import('./db')
    runInAction(() => {
      this.dbHospedeiros = hospedeiros
      this.dbRegras = regras
      this.dbPragas = pragas
      this.estados = estados
      this.db = this.dbRegras.map(regra => {
        const praga = this.dbPragas.find(item => item.prag === regra.prag)
        if (!praga) {
          throw Error(`Dados da praga ${regra.prag} nÃ£o cadastrados.`)
        } else {
          return { ...regra, ...praga }
        }
      })
    })
  }

  @computed get hospedeirosPragas() {
    return this.dbPragas.flatMap(praga => praga.hosp)
  }

  @computed get hospedeirosRegulamentados() {
    return this.dbHospedeiros.filter(hospedeiro =>
      this.hospedeirosPragas.includes(hospedeiro.nomeSci)
    )
  }

  @computed get listaNomesSci() {
    return this.hospedeirosRegulamentados.unique('nomeSci').sort((a, b) => a.localeCompare(b))
  }

  @computed get listaNomesVul() {
    return this.hospedeirosRegulamentados.unique('nomeVul').sort((a, b) => a.localeCompare(b))
  }

  @computed
  get empty(): boolean {
    return this.result.length === 0
  }

  @computed
  get origem() {
    return this.estados.filter(estado => estado.UF !== this.dados.dest || estado.UF === '')
  }

  @computed
  get destino() {
    return this.estados.filter(estado => estado.UF !== this.dados.orig || estado.UF === '')
  }

  @computed
  get gender(): string {
    return this.dados.hospSci.split(' ')[0]
  }

  @computed
  get completed(): boolean {
    return (
      Boolean(this.dados.hospSci) &&
      Boolean(this.dados.hospVul) &&
      Boolean(this.dados.prod) &&
      Boolean(this.dados.orig) &&
      Boolean(this.dados.dest)
    )
  }

  @computed
  get partes(): string[] {
    return this.db
      .filter(
        exigen =>
          exigen.hosp.includes(this.dados.hospSci) ||
          exigen.hosp.includes(`${this.gender} sp.`) ||
          exigen.hosp.includes(`${this.gender} spp.`)
      )
      .by('part')
      .flat()
      .unique()
      .concat([''])
      .sort((a: string, b: string) => a.localeCompare(b))
  }

  @computed
  get result() {
    return this.db.filter(exigen => {
      return (
        (exigen.hosp.includes(this.dados.hospSci) ||
          exigen.hosp.includes(`${this.gender} sp.`) ||
          exigen.hosp.includes(`${this.gender} spp.`)) &&
        exigen.orig.includes(this.dados.orig) &&
        exigen.dest.includes(this.dados.dest) &&
        exigen.part.includes(this.dados.prod)
      )
    })
  }

  @action
  handleChanges = (event: { target: { name: string; value: string } }): void => {
    switch (event.target.name) {
      case 'hospSci':
        const hospVulg = this.dbHospedeiros.find(hosp => hosp.nomeSci === event.target.value)
        this.dados.hospVul = hospVulg ? hospVulg.nomeVul : ''
        break
      case 'hospVul':
        const hospSci = this.dbHospedeiros.find(hosp => hosp.nomeVul === event.target.value)
        this.dados.hospSci = hospSci ? hospSci.nomeSci : ''
        break
      default:
        break
    }
    this.dados[event.target.name] = event.target.value
  }

  @action
  clean(): void {
    this.dados.hospSci = ''
    this.dados.hospVul = ''
    this.dados.prod = ''
    this.dados.orig = ''
    this.dados.dest = ''
  }
}

const store = new Store()
store.getDb()
export { store }
