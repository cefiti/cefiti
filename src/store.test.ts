import { store } from './store'
//import { configure } from 'mobx'
import { regras } from './dbRegras'
import { pragas } from './dbPragas'
import { hospedeiros } from './dbHospedeiros'
//import '../assests/utils'
import * as d3 from 'd3-array'
//import 'jest'

//configure({ enforceActions: 'observed' }) //useStrict(true)

const estadosSemAC = [
  { estado: '', UF: '' },
  //{estado: 'Acre', UF: 'AC'},
  { estado: 'Alagoas', UF: 'AL' },
  { estado: 'Amazonas', UF: 'AM' },
  { estado: 'Amapá', UF: 'AP' },
  { estado: 'Bahia', UF: 'BA' },
  { estado: 'Ceará', UF: 'CE' },
  { estado: 'Distrito Federal', UF: 'DF' },
  { estado: 'Espirito Santo', UF: 'ES' },
  { estado: 'Goiás', UF: 'GO' },
  { estado: 'Maranhão', UF: 'MA' },
  { estado: 'Minas Gerais', UF: 'MG' },
  { estado: 'Mato Grosso do Sul', UF: 'MS' },
  { estado: 'Mato Grosso', UF: 'MT' },
  { estado: 'Pará', UF: 'PA' },
  { estado: 'Paraíba', UF: 'PB' },
  { estado: 'Pernambuco', UF: 'PE' },
  { estado: 'Piauí', UF: 'PI' },
  { estado: 'Paraná', UF: 'PR' },
  { estado: 'Rio de janeiro', UF: 'RJ' },
  { estado: 'Rio Grande do Norte', UF: 'RN' },
  { estado: 'Rondônia', UF: 'RO' },
  { estado: 'Roraima', UF: 'RR' },
  { estado: 'Rio Grande do Sul', UF: 'RS' },
  { estado: 'Santa Catarina', UF: 'SC' },
  { estado: 'Sergipe', UF: 'SE' },
  { estado: 'São Paulo', UF: 'SP' },
  { estado: 'Tocantins', UF: 'TO' },
]

describe('Store origem e destino', () => {
  it('origem AC', () => {
    store.dados.orig = 'AC'
    expect(store.destino).toEqual(estadosSemAC)
  })

  it('origem MT', () => {
    store.dados.orig = 'MT'
    expect(store.destino).not.toEqual(estadosSemAC)
  })

  it('destino AC', () => {
    store.dados.dest = 'AC'
    expect(store.origem).toEqual(estadosSemAC)
  })

  it('destino MT', () => {
    store.dados.dest = 'MT'
    expect(store.origem).not.toEqual(estadosSemAC)
  })
})

describe('Store hospedeiros nomeSci', () => {
  it('unique values Nome Vulgar', () => {
    //expect(store.listaNomesVul.length).toEqual(hospedeiros.length)
    expect(hospedeiros.map((v) => v.nomeVul).filter((i, x, a) => a.indexOf(i) !== x)).toEqual([])
    expect(
      hospedeiros.map((v) => v.nomeVul).filter((i, x, a) => a.indexOf(i) === x).length
    ).toEqual(hospedeiros.length)
  })
})

describe('Store partes', () => {
  it('de Acerola', () => {
    debugger
    store.dados.hospSci = 'Malpighia spp.'
    expect(store.partes).toEqual(['', 'frutos'])
  })

  it('de Banana', () => {
    store.dados.hospSci = 'Musa spp.'
    expect(store.partes).toEqual([
      '',
      'flores',
      'frutos',
      'material para pesquisa',
      'mudas',
      'rizomas',
    ])
  })

  it('de Citrus', () => {
    store.dados.hospSci = 'Citrus spp.'
    expect(store.partes).toEqual([
      '',
      'caules',
      'estacas',
      'flores',
      'folhas',
      'frutos',
      'gemas',
      'material de propagação',
      'mudas',
      'plantas',
      'raízes',
      'ramas',
      'ramos',
    ])
  })

  /* it('de Mandioca', () => {
    store.dados.hospSci = 'Manihot esculenta'
    expect(store.partes).toEqual(['', 'estaca', 'madeira', 'maniva'])
  }) */
})

describe('Store: gender', () => {
  it('Acacia sp.', () => {
    store.dados.hospSci = 'Acacia sp.'
    expect(store.gender).toBe('Acacia')
  })

  it('Acer macrophyllum', () => {
    store.dados.hospSci = 'Acer macrophyllum'
    expect(store.gender).toBe('Acer')
  })

  it('Betula lutea (synonym: alleghaniensis) ', () => {
    store.dados.hospSci = 'Betula lutea (synonym: alleghaniensis)'
    expect(store.gender).toBe('Betula')
  })
})

describe('Store filtro geral', () => {
  beforeAll(() => {
    store.dados.hospSci = 'Musa spp.'
    store.dados.prod = 'frutos'
    store.dados.orig = 'MG'
    store.dados.dest = 'MT'
  })

  it('Musa spp. count', () => {
    expect(store.result.length).toBe(3)
  })
  it('Musa spp. legis', () => {
    expect(store.result.flatMap((v) => v.files).map((v) => v.link)).toEqual([
      'IN17-2009.pdf',
      'IN17-2005.pdf',
      'IN17-2005.pdf',
    ])
  })

  it('Musa spp. pragas', () => {
    expect(store.result.map((v) => v.pragc)).toEqual([
      'MOKO-DA-BANANEIRA',
      'SIGATOKA NEGRA',
      'SIGATOKA NEGRA',
    ])
  })

  it('Eugenia uniflora', () => {
    store.dados.hospSci = 'Eugenia uniflora'
    store.dados.prod = 'frutos'
    store.dados.orig = 'PI'
    store.dados.dest = 'DF'
    expect(store.result.length).toBe(1)
  })

  it('Malus spp.', () => {
    store.dados.hospSci = 'Malus spp.'
    store.dados.prod = 'frutos'
    store.dados.orig = 'SC'
    store.dados.dest = 'MT'
    expect(store.result.length).toBe(1)
    expect(store.result.flatMap((v) => v.files).map((v) => v.link)).toEqual(['IN20-2013.pdf'])
    expect(store.result.map((v) => v.pragc)).toEqual(['CANCRO EUROPEU DAS POMÁCEAS'])
  })

  it('Citrus sinensis sementes SP->ES', () => {
    store.dados.hospSci = 'Citrus sinensis'
    store.dados.prod = 'sementes'
    store.dados.orig = 'SP'
    store.dados.dest = 'ES'
    expect(store.result.length).toBe(0)
    //expect(store.result.by('files').flatten().by('link')).toEqual(['PORT291-1997.pdf', 'IN53-2008.pdf'])
    //expect(store.result.by('pragc')).toEqual(['CANCRO CÍTRICO', 'GREENING'])
  })

  it('Citrus sinensis sementes RS->ES', () => {
    store.dados.hospSci = 'Citrus sinensis'
    store.dados.prod = 'material de propagação'
    store.dados.orig = 'RS'
    store.dados.dest = 'ES'
    expect(store.result.length).toBe(2)
    expect(store.result.flatMap((v) => v.files).map((v) => v.link)).toEqual([
      'IN03-2008.pdf',
      'IN21-2018.pdf',
    ])
    expect(store.result.map((v) => v.pragc)).toEqual(['PINTA-PRETA-DOS-CITROS', 'CANCRO CÍTRICO'])
  })

  it('Citrus sinensis mudas SP->ES', () => {
    store.dados.hospSci = 'Citrus sinensis'
    store.dados.prod = 'mudas'
    store.dados.orig = 'SP'
    store.dados.dest = 'ES'
    expect(store.result.length).toBe(3)
    expect(store.result.flatMap((v) => v.files).map((v) => v.link)).toEqual([
      'IN53-2008.pdf',
      'IN03-2008.pdf',
      'IN21-2018.pdf',
    ])
    expect(store.result.map((v) => v.pragc)).toEqual([
      'GREENING',
      'PINTA-PRETA-DOS-CITROS',
      'CANCRO CÍTRICO',
    ])
    expect(store.result).toMatchSnapshot()
  })
})

describe('Sync between NomeVulg and NomeSci', () => {
  it('should define NomeVulg based in NomeSci', () => {
    // @ts-ignore
    const e: EventChange = { currentTarget: { name: 'hospSci', value: 'Musa spp.' } }
    store.handleChanges(e)
    //store.dados.hospSci = 'Musa spp.'
    expect(store.dados.hospVul).toEqual('Banana')
  })
  it('should define NomeSci  based in NomeVulg ', () => {
    // @ts-ignore
    store.handleChanges({ currentTarget: { name: 'hospVul', value: 'Banana' } })
    //store.dados.hospVul = 'Banana'
    expect(store.dados.hospSci).toEqual('Musa spp.')
  })
})

test('Check normalization of db ', () => {
  regras.map((regra) => {
    const praga = pragas.find((item) => item.prag === regra.prag)
    if (!praga) {
      expect(regra.prag).toEqual(praga)
      //throw Error(`Dados da praga ${regra.prag} não cadastrados.`)
    }
    return regra
  })
})

test('duplicates nomeVul', () => {
  const countDupli = Array.from(
    d3.rollup(
      hospedeiros,
      (v) => ({ countNomeVulg: v.length }),
      (k) => k.nomeVul
    ),
    ([key, values]) => ({ nomeVulg: key, ...values })
  )
  expect(countDupli.filter((v) => !v.countNomeVulg)).toEqual([])
})

test('should join Pragas and Regras', () => {
  regras.forEach((regra) => {
    const praga = pragas.find((item) => item.prag === regra.prag)
    expect(praga).toBeDefined()
  })
})
