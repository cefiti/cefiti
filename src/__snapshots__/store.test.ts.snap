// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store filtro geral Citrus sinensis mudas SP->ES 1`] = `
Array [
  Object {
    "desc": "DE UF com ocorrência de Greening PARA UF reconhecida pelo MAPA como livre da ocorrência de Greening",
    "dest": Array [
      "AC",
      "AL",
      "AM",
      "AP",
      "BA",
      "CE",
      "DF",
      "ES",
      "GO",
      "MA",
      "MS",
      "MT",
      "PA",
      "PB",
      "PE",
      "PI",
      "RJ",
      "RN",
      "RO",
      "RR",
      "RS",
      "SC",
      "SE",
      "TO",
    ],
    "exig": Array [
      "PTV; 
 Obs.: Partida apreendida pela fiscalização de defesa sanitária vegetal, em
desacordo com o previsto nesta Instrução Normativa, será sumariamente destruída, não cabendo ao infrator qualquer tipo de indenização, sem prejuízo das demais sanções estabelecidas pela legislação estadual e federal de defesa sanitária vegetal, conforme o art. 5º, da Instrução Normativa MAPA nº. 53, de 16 de outubro de 2010.",
    ],
    "files": Array [
      Object {
        "leg": "Instrução Normativa nº 53, de 16 de outubro de 2008",
        "link": "IN53-2008.pdf",
      },
    ],
    "hosp": Array [
      "Citrus spp.",
      "Fortunella spp.",
      "Poncirus spp.",
      "Murraya paniculata",
    ],
    "orig": Array [
      "MG",
      "PR",
      "SP",
    ],
    "part": Array [
      "mudas",
      "estacas",
      "gemas",
      "ramas",
      "raízes",
      "material de propagação",
      "plantas",
    ],
    "prag": "Candidatus liberibacter americanus e Candidatus liberibacter asiaticus",
    "pragc": "GREENING",
  },
  Object {
    "desc": "DE Área de Sistema de Manejo de Risco (SMR) de Cancro Cítrico PARA qualquer UF",
    "dest": Array [
      "AC",
      "AL",
      "AM",
      "AP",
      "BA",
      "CE",
      "DF",
      "ES",
      "GO",
      "MA",
      "MG",
      "MS",
      "MT",
      "PA",
      "PB",
      "PE",
      "PI",
      "PR",
      "RJ",
      "RN",
      "RO",
      "RR",
      "RS",
      "SC",
      "SE",
      "SP",
      "TO",
    ],
    "exig": Array [
      "PTV com a seguinte DA: \\"O material de propagação é proveniente de Área sob SMR, se encontra livre de Cancro Cítrico (Xanthomonas citri subsp. citri) e foi produzido conforme preconiza a legislação específica em vigor\\".",
    ],
    "files": Array [
      Object {
        "leg": "Instrução Normativa nº 21, de 25 de abril de 2018",
        "link": "IN21-2018.pdf",
      },
    ],
    "hosp": Array [
      "Citrus spp.",
      "Fortunella spp.",
      "Poncirus spp.",
    ],
    "orig": Array [
      "MT",
      "MS",
      "RS",
      "SP",
      "SC",
    ],
    "part": Array [
      "material de propagação",
      "mudas",
      "estacas",
      "gemas",
      "plantas",
    ],
    "prag": "Xanthomonas citri subsp. citri",
    "pragc": "CANCRO CÍTRICO",
  },
]
`;
